cli/internal/oauth/manager/manager_test.go

```diff
@@ -0,0 +1,277 @@
+	validParsedToken = oauth.TokenResult{
+		AccessToken:  validToken,
+		RefreshToken: "refresh-token",
+		Claims: oauth.Claims{
+			Claims: jwt.Claims{
+				Issuer:  "https://login.docker.com/",
+				Subject: "samlp|samlp-docker|bork@docker.com",
+				Audience: jwt.Audience{
+					"https://audience.com",
+				},
+				Expiry:   &expiry,
+				IssuedAt: &issuedAt,
+			},
+			Domain: oauth.DomainClaims{
+				UUID:      "0123-456789",
+				Email:     "bork@docker.com",
+				Username:  "bork!",
+				Source:    "samlp",
+				SessionID: "a-session-id",
+			},
+			Scope: "openid offline_access",
+		},
+	}
+)
+
+func TestLoginDevice(t *testing.T) {
+	t.Run("valid token", func(t *testing.T) {
+		expectedState := api.State{
+			DeviceCode:      "device-code",
+			UserCode:        "0123-4567",
+			VerificationURI: "an-url",
+			ExpiresIn:       300,
+		}
+		var receivedAudience string
+		getDeviceToken := func(audience string) (api.State, error) {
+			receivedAudience = audience
+			return expectedState, nil
+		}
+		var receivedState api.State
+		waitForDeviceToken := func(state api.State) (api.TokenResponse, error) {
+			receivedState = state
+			return api.TokenResponse{
+				AccessToken:  validToken,
+				RefreshToken: "refresh-token",
+			}, nil
+		}
+		api := &testAPI{
+			getDeviceToken:     getDeviceToken,
+			waitForDeviceToken: waitForDeviceToken,
+		}
+		manager := OAuthManager{
+			audience: "https://hub.docker.com",
+			api:      api,
+			openBrowser: func(url string) error {
+				return nil
+			},
+		}
+
+		res, err := manager.LoginDevice(context.Background(), os.Stderr)
+		assert.NilError(t, err)
+
+		assert.Equal(t, receivedAudience, "https://hub.docker.com")
+		assert.Equal(t, receivedState, expectedState)
+		assert.DeepEqual(t, *res, validParsedToken)
+	})
+
+	t.Run("timeout", func(t *testing.T) {
+		getDeviceToken := func(audience string) (api.State, error) {
+			return api.State{
+				DeviceCode:      "device-code",
+				UserCode:        "0123-4567",
+				VerificationURI: "an-url",
+				ExpiresIn:       300,
+			}, nil
+		}
+		waitForDeviceToken := func(state api.State) (api.TokenResponse, error) {
+			return api.TokenResponse{}, api.ErrTimeout
+		}
+		a := &testAPI{
+			getDeviceToken:     getDeviceToken,
+			waitForDeviceToken: waitForDeviceToken,
+		}
+		manager := OAuthManager{
+			api: a,
+			openBrowser: func(url string) error {
+				return nil
+			},
+		}
+
+		_, err := manager.LoginDevice(context.Background(), os.Stderr)
+		assert.ErrorContains(t, err, "login failed: timed out waiting for device token")
+	})
+
+	t.Run("canceled context", func(t *testing.T) {
+		getDeviceToken := func(audience string) (api.State, error) {
+			return api.State{
+				DeviceCode: "device-code",
+				UserCode:   "0123-4567",
+			}, nil
+		}
+		waitForDeviceToken := func(state api.State) (api.TokenResponse, error) {
+			// make sure that the context is cancelled before this returns
+			time.Sleep(500 * time.Millisecond)
```

Comment:
Possibly we could move the `ctx, cancel := context.WithCancel(context.Background())` earlier and then just:
```suggestion
            select {
			case <-ctx.Done():
			case <-time.After(time.Second):
			}
```

so we don't need to always wait for the 500ms?

Reply:
The issue is in https://github.com/laurazard/cli/blob/6a838e2625c2e4246e76dbd6fbda52634917e5fa/cli/internal/oauth/manager/manager.go#L98-L104 – if both channels have an item there, `select` will take one at random. Cancelling earlier doesn't fix it, we just need to make sure it returns after (or never would be an option too).

Reply:
maybe the comment wasn't clearer, it's not to ensure the context is cancelled, it's to ensure that the `select` statement selects from the context.Done channel and not from the "result" channel.

Reply:
Ah okay! Also looks like it will be called in a separate goroutine anyway, so it won't stall the main test goroutine, so I'm fine 👍🏻